core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1103,35 +1103,41 @@ class KafkaApis(val requestChannel: RequestChannel,
 
     val responseTopics = authorizedRequestInfo.map { topic =>
       val responsePartitions = topic.partitions.asScala.map { partition =>
-        val topicPartition = new TopicPartition(topic.name, partition.partitionIndex)
-
-        try {
-          val offsets = replicaManager.legacyFetchOffsetsForTimestamp(
-            topicPartition = topicPartition,
-            timestamp = partition.timestamp,
-            maxNumOffsets = partition.maxNumOffsets,
-            isFromConsumer = offsetRequest.replicaId == ListOffsetsRequest.CONSUMER_REPLICA_ID,
-            fetchOnlyFromLeader = offsetRequest.replicaId != ListOffsetsRequest.DEBUGGING_REPLICA_ID)
+        if (partition.timestamp() < ListOffsetsRequest.EARLIEST_TIMESTAMP) {
           new ListOffsetsPartitionResponse()
             .setPartitionIndex(partition.partitionIndex)
-            .setErrorCode(Errors.NONE.code)
-            .setOldStyleOffsets(offsets.map(JLong.valueOf).asJava)
-        } catch {
-          // NOTE: UnknownTopicOrPartitionException and NotLeaderOrFollowerException are special cases since these error messages
-          // are typically transient and there is no value in logging the entire stack trace for the same
-          case e @ (_ : UnknownTopicOrPartitionException |
-                    _ : NotLeaderOrFollowerException |
-                    _ : KafkaStorageException) =>
-            debug("Offset request with correlation id %d from client %s on partition %s failed due to %s".format(
-              correlationId, clientId, topicPartition, e.getMessage))
-            new ListOffsetsPartitionResponse()
-              .setPartitionIndex(partition.partitionIndex)
-              .setErrorCode(Errors.forException(e).code)
-          case e: Throwable =>
-            error("Error while responding to offset request", e)
+            .setErrorCode(Errors.UNSUPPORTED_VERSION.code)
+        } else {
+          val topicPartition = new TopicPartition(topic.name, partition.partitionIndex)
+
+          try {
+            val offsets = replicaManager.legacyFetchOffsetsForTimestamp(
```

Comment:
This is another interesting issue. 

#11390 added the support of `EARLIEST_LOCAL_TIMESTAMP` to v0 path. That seems to make server-to-server replica sync path work (LeaderEndPoint#fetchEarliestLocalOffset). However, the LIST_OFFSET request sent by server will use version_0 only if  the MV is smaller than `0.10.2`. What is the scenario that we set MV be smaller than `0.10.2` - the cluster has some brokers running by distribution < 0.10.1

However, those older brokers can't ever handle `EARLIEST_LOCAL_TIMESTAMP` correctly as they have no idea of `EARLIEST_LOCAL_TIMESTAMP`.

IMHO, we should revert the support of `EARLIEST_LOCAL_TIMESTAMP` from v0 path. Also, `LeaderEndPoint#fetchEarliestLocalOffset` should require the min version >=8

@jolshan @FrankYang0529 @junrao WDYT?

Reply:
> IMHO, we should revert the support of EARLIEST_LOCAL_TIMESTAMP from v0 path. 

Sounds good to me.

> LeaderEndPoint#fetchEarliestLocalOffset should require the min version >=8

LeaderEndPoint#fetchEarliestLocalOffset is only used in tier storage, which is released in 3.6.0. An MV of 3.6.0 implies that the ListOffset request will use version >=8. So, it seems this is already covered?

Reply:
> LeaderEndPoint#fetchEarliestLocalOffset is only used in tier storage, which is released in 3.6.0. An MV of 3.6.0 implies that the ListOffset request will use version >=8. So, it seems this is already covered?

Still trying to understand everything here, but just because it was released 3.6, doesn't mean MV is 3.6 unless tiered storage requires that as minimum MV to enable? If so, then I agree this is covered, but maybe should include a comment so it is not used for other things.

Reply:
I do see this: https://github.com/apache/kafka/blob/0bb2aee8385c98d38b306f77a8c71875f5d48c1a/server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java#L477

Reply:
>  Still trying to understand everything here, but just because it was released 3.6, doesn't mean MV is 3.6 unless tiered storage requires that as minimum MV to enable? If so, then I agree this is covered, but maybe should include a comment so it is not used for other things.

I agree to @jolshan. Also, we can't guarantee the method is used by tiered storage always. At least, we should add comments to that method to reminder us to use it carefully.

> 3.5-IV0 marked as the MV for tiered storage it seems

https://github.com/apache/kafka/blob/trunk/server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java#L171

Reply:
I saw this but wasn't sure where it was enforced -- besides the apis tiered storage uses. (which maybe is enough?) I expected a broker to fail to set the config if the MV was too low or something.

Reply:
> I saw this but wasn't sure where it was enforced -- besides the apis tiered storage uses. (which maybe is enough?) I expected a broker to fail to set the config if the MV was too low or something.

yep, apis-level check should be good enough. My point was that broker should set "correct" min version for both `LIST_OFFSET` and `FETCH` according to MV. It makes broker receive the `INVALID_REQUEST` error which is better than "incorrect" response created by (older) brokers

Reply:
Currently, it doesn't seem there is enforcement of at least MV 3.5 when enabling tier storage. It's probably useful to add one. What do you think, @satishd ?

Reply:
> My point was that broker should set "correct" min version for both LIST_OFFSET and FETCH according to MV.

Feature version now supports MV dependency. Going forward, instead of enforcing a minimum MV at each RPC level, it's probably better to enforce it once at the feature version level.

Reply:
If tiered storage can have correct MV, enforcing a minimum MV at each RPC level is not necessary as the request version is based on MV. 

Hence, the issue is "how to enforce the MV 3.5 when enabling tiered storage"?

> Feature version now supports MV dependency. Going forward, instead of enforcing a minimum MV at each RPC level, it's probably better to enforce it once at the feature version level.

Pardon me, do you mean to add `TieredStorageVersion` (like `GroupVersion`)? if so, I guess that is beyond 3.9? 

IMHO, checking the version when tiered storage sends request is a enough simple solution to backport to 3.9.

Reply:
Yeah, I think since it is already like this, it doesn't make sense to make a drastic change for 3.9

Reply:
In order to be a good friend to 3.9, I will file jira to trace the possible drastic change we discussed.

**SHOULD** be included by this PR

1. revert the support of EARLIEST_LOCAL_TIMESTAMP from v0 path. see comment: https://github.com/apache/kafka/pull/16873#discussion_r1725714630
2. adjust version of `LIST_OFFSET`. the version must be >= 8  see code:
https://github.com/apache/kafka/blob/61a661ec5e627217e8b4e4c009d65ee0e0e938ba/core/src/main/scala/kafka/server/RemoteLeaderEndPoint.scala#L107
This can protect broker from sending request `EARLIEST_LOCAL_TIMESTAMP` to the old brokers which handle `EARLIEST_LOCAL_TIMESTAMP` as normal TS.

**SHOULD NOT** be included by this PR
1. adjust version of `FETCH`. the version=14 adds new error handle for `OffsetMovedToTieredStorageException` only. Hence, it is ok to send request to old brokers as they don't have tiered storage yet.
2. enforce Tiered storage once at the feature version level. this will be traced by https://issues.apache.org/jira/browse/KAFKA-17405


@FrankYang0529 @junrao @jolshan PTAL

Reply:
> IMHO, checking the version when tiered storage sends request is a enough simple solution to backport to 3.9.

This is simple, but I am not sure if this completely solves the problem. Note that EARLIEST_LOCAL_TIMESTAMP is not always used by tier storage. So, if someone enables tier storage with the wrong MV, the ListOffset issue may show up at a much later time. If the broker has the wrong MV, it's probably better to prevent enabling tier storage in the first place.

For 3.9, I was thinking that we could change `createRemoteLogManager()` in BrokerServer to take the MV into consideration. If the MV is too low, we could just return None even if remote storage has been enabled. This is a bit tricky since MV is only available after the following statement.

```
      FutureUtils.waitWithLogging(logger.underlying, logIdent,
        "the initial broker metadata update to be published",
        brokerMetadataPublisher.firstPublishFuture , startupDeadline, time)
```
However, remoteLogManager is created and passed into ReplicaManager before that. Maybe we could have a method to set the remoteLogManager instead of passing it through the constructor.

Reply:
@junrao -- this makes sense, but want to confirm -- has this been an issue since tiered storage was released?

Reply:
I am not sure if people actually hit this issue. @satishd : Do you know?

Reply:
> For 3.9, I was thinking that we could change createRemoteLogManager() in BrokerServer to take the MV into consideration. If the MV is too low, we could just return None even if remote storage has been enabled. This is a bit tricky since MV is only available after the following statement.

That is indeed a good protection, but as we discussed in another PR - the MV is dynamic so we still need to check the version of request sent by tiered storage in runtime, right?

I guess the total solution (make tiered storage work with dynamic MV) is "feature version", but the "acceptable" fixes we can address for now are:

1. disable tiered storage when MV is <= 3.5 in startup. This CAN protect users from using tiered storage with incorrect MV in the beginning, but it CAN NOT work with dynamic MV. i.e tiered storage CAN NOT be re-enabeld/re-disabled when MV gets upgrade/downgrade
2. produce error when tiered storage sends request with incorrect MV. This CAN make "noise" to tiered storage when MV gets upgrade/downgrade, but the error CAN NOT be there in time. That means the error happens only if tiered storage use the request.

> this makes sense, but want to confirm -- has this been an issue since tiered storage was released?

The issue we discuss happens on LIST_OFFSET request, and it is used to deal with `OFFSET_MOVED_TO_TIERED_STORAGE`. The error is produced after 3.6 (https://github.com/apache/kafka/commit/6f197301646135e0bb39a461ca0a07c09c3185fb#diff-78812e247ffeae6f8c49b1b22506434701b1e1bafe7f92ef8f8708059e292bf0). That means the leader (broker) which can return the `OFFSET_MOVED_TO_TIERED_STORAGE` is able to handle `EARLIEST_LOCAL_TIMESTAMP` too. 

So my first comment may be no correct: ` those older brokers can't ever handle EARLIEST_LOCAL_TIMESTAMP correctly as they have no idea of EARLIEST_LOCAL_TIMESTAMP`.

In short, we can let it go if all we want (know) to fix is the LIST_OFFSET request. Or we can add some protection (as above).

Reply:
I do not remember users hitting this issue. It is required to be mandatory to upgrade to 3.6 version and set the respective IBP before enabling tiered storage at the cluster level.

Reply:
Can you point me where this validation is done? I looked for it but must have missed it.

Reply:
Just to be clear, validation is missing here. However, users enable it only when they upgrade to 3.6 and set the respective IBP. 

+1 to add the validation to protect users in misconfiguring, it is good to add that validation in 3.9.

Reply:
> +1 to add the validation to protect users in misconfiguring, it is good to add that validation in 3.9.

@satishd thanks for confirmation. I'd like to keep discussing the validation for MV and remote storage.


*zk broker*: the MV is static so we can check the MV when creating RLM.

*kraft broker*: the MV is dynamic, and we all agree that RLM needs respective IBP (3.6). Maybe we can set the broker's min feature version of MV from `IBP_3_0_IV1` to `IBP_3_6_IV0` when RLM is enabled. That means the broker with RLM can't join the cluster which does not support the 3.6+. This changes also have following benefit. 

1) we don't need to handle tricky MV sync ( https://github.com/apache/kafka/pull/16873#discussion_r1728047972)
2) we don't need to worry the MV gets downgrade to 3.6-

@junrao @satishd @jolshan @FrankYang0529 WDYT? 

BTW, we have a ticker for the validation (https://issues.apache.org/jira/browse/KAFKA-17405), but we can reach the consensus here. I will copy the solution to the jira. In the mean time, @FrankYang0529 maybe you can address the `LIST_OFFSET`-related check in this PR (include this comment: https://github.com/apache/kafka/pull/16873#discussion_r1725714630)

Reply:
Regarding gating tier storage on MV 3.6, we could potentially pass in a remoteLogManagerSupplier instead of a direct remoteLogManager to ReplicaManager. The supplier will instantiate remoteLogManager on first usage based on the remote storage config and the MV setting at that time. The MV initialization happens before the opening of socket server. So, by the time the supplier is called, we can be sure that MV has been initialized. This still doesn't support enabling remote storage dynamically, but is probably good enough in 3.9.

Reply:
@junrao with your approach, what happens if the MV gets downgraded? 

I also wasn't sure if we supported downgrades of MV at this time. 3.6-IV0 doesn't change metadata so in theory it is safe to downgrade to 3.5-IV2

Reply:
I am not sure if tier storage supports downgrading MV. @satishd Do you know?

Reply:
> This still doesn't support enabling remote storage dynamically, but is probably good enough in 3.9.

enabling remote storage is another issue, and I agree to your approach is good and simple in 3.9

 > what happens if the MV gets downgraded?

that is a good question. In the request-level, both version=14 of `FETCH` and version=8 of `LIST_OFFSET` are supported by 3.5-IV2. Hence, it should be fine even though MV gets safe-downgraded to 3.5-IV2.

> I am not sure if tier storage supports downgrading MV

I guess that will be a issue in the future if tier storage requires more and more new requests to address more features.

IMHO we can adopt @junrao's approach - checking the cluster MV (3.6) when creating tier storage. The side effect is that we will "allow" safe-downgrade to 3.5-IV2 even though tiered storage is enabled. That is a bit weird since we declare tier storage is early access feature available as of 3.6.

Reply:
>I am not sure if tier storage supports downgrading MV. @satishd Do you know?

@junrao Are you asking about downgrading it to 3.5-IV2? If so, it seems to be safe to downgrade it to 3.5 IV2 as the required APIs are available with that version.

Reply:
> If so, it seems to be safe to downgrade it to 3.5 IV2 as the required APIs are available with that version.

Does tier storage work in 3.5?

Reply:
> Does tier storage work in 3.5?

*RPC*: both version=14 of FETCH and version=8 of LIST_OFFSET are included by 3.5-IV2
*config*:  tier storage configs are "internal" https://github.com/apache/kafka/blob/3.5/storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java#L128
*production-ready": tier storage is "preview" even in 3.6 https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Tiered+Storage+Early+Access+Release+Notes

In short, users should NOT run tier storage with 3.5 binary, but related RPCs are included by 3.5 - they may be not full-implemented but the RPCs do not return error of unsupported version.

Reply:
Sorry for the confusion with my earlier comment. I was referring to the IBP downgrade, not the binaries. Regardless, this isn't intended for production use. With version 3.6, users are advised to experiment with tiered storage only on development clusters, not in production environments, as mentioned [here](https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Tiered+Storage+Early+Access+Release+Notes).

Reply:
Sorry for the confusion with my earlier comment. I was referring to the IBP downgrade, not the binaries. Regardless, this isn't intended for production use. With version 3.6, users are suggested to experiment with tiered storage only on development clusters, not in production environments, as mentioned [here](https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Tiered+Storage+Early+Access+Release+Notes).

Reply:
Thanks for the explanation, Satish. Since tier storage is EA-ed in 3.6, it makes sense to require MV 3.6 to turn on tier storage. This does mean that tier storage won't work with MV 3.5 and a binary on 3.6 or later. But that's probably ok.

Reply:
Dear all, do we reach the consensus on any approach to validate MV for tier storage?

1. RPC version check: works with dynamical MV but cannot do fail-fast

2. RLM initialization check: good at fail-fast but cannot check dynamic MV after initialization. BTW, there was an issue (https://issues.apache.org/jira/browse/KAFKA-16790) used to change the initialization order of RLM and metadata publisher. 

3. Set min MV to 3.6 when RLM is enabled. Good at fail-fast and dynamic MV check. However, it is a bit weird to change the min MV by the non-feature-version.

4. Ship tier storage to feature version? Too large to be in 3.9

5. Let it go. We have announced the requirement of tier storage, so users should be aware of that!

Reply:
Thanks @chia7712 for the summary. I'll go with (1) or (3).

Reply:
1 as implemented in this PR doesn't do what's expected, right? See https://github.com/apache/kafka/pull/16873#discussion_r1731530754

Reply:
> 1 as implemented in this PR doesn't do what's expected

I assumed the check happens in creating the `LIST_OFFSET`/`FETCH` request. for example:

```java
  override def fetchEarliestLocalOffset(topicPartition: TopicPartition, currentLeaderEpoch: Int): OffsetAndEpoch = {
    val metadataVersion = metadataVersionSupplier()
    if (!metadataVersion.isEarliestLocalOffsetSupported)
      throw new UnsupportedVersionException("Metadata Version is unmatched. It requires 3.5 but current is " + metadataVersion)
    fetchOffset(topicPartition, currentLeaderEpoch, ListOffsetsRequest.EARLIEST_LOCAL_TIMESTAMP)
  }
```

Reply:
Hi @junrao and @chia7712, thanks for the suggestion and review. I reverted the change in `ListOffsetsRequest` and  added check to `RemoteLeaderEndPoint#fetchEarliestLocalOffset`.

Reply:
Dear all, the validation to MV is not regression and we don't reach consensus before 3.9 code freeze. IMHO, we should move the validation discussion to https://issues.apache.org/jira/browse/KAFKA-17405 and than @FrankYang0529 can focus on the validation of `LIST_OFFSET` which returns unsupported version if the timestamp flag is used by the incorrect version